packages/ui/src/components/Tooltip.tsx

```diff
@@ -0,0 +1,332 @@
+        },
+        trianglePosition: {},
+      };
+    case "inset-bottom-left":
+      return {
+        tooltipPosition: { bottom: 0, left: 0 },
+        trianglePosition: {},
+      };
+    case "inset-top-left":
+      return {
+        tooltipPosition: { top: 0, left: 0 },
+        trianglePosition: {},
+      };
+    case "inset-top-right":
+      return {
+        tooltipPosition: { top: 0, right: 0 },
+        trianglePosition: {},
+      };
+    case "inset-bottom-right":
+      return {
+        tooltipPosition: { bottom: 0, right: 0 },
+        trianglePosition: {},
+      };
+    default:
+      return { tooltipPosition: {}, trianglePosition: {} };
+  }
+};
+
+type UIPosition = {
+  top?: number;
+  bottom?: number;
+  left?: number;
+  right?: number;
+  width?: number;
+  height?: number;
+};
+
+type TooltipState = {
+  tooltipPosition?: UIPosition;
+  trianglePosition?: UIPosition;
+};
+
+type Position = {
+  x: number;
+  y: number;
+};
+
+type Size = {
+  width: number;
+  height: number;
+};
+
+export default (props: TooltipProps) => {
+  const { placement = "top", show, anchorRef, children } = props;
+
+  const tooltipRef = useRef<View | null>(null);
+  const containerRef = useRef<View | null>(null);
+
+  const [{ tooltipPosition, trianglePosition }, dispatch] = useReducer(
+    (
+      state: TooltipState,
+      { type, payload }: { type: string; payload: any },
+    ) => {
+      if (type === "init") {
+        return { ...getPositions(payload) };
+      }
+      if (type === "move") {
+        return {
+          tooltipPosition: { ...payload },
+        };
+      }
+      return state;
+    },
+    {},
+  );
+
+  const [containerPosition, setContainerPosition] = useState<Position | null>(
+    null,
+  );
+  useEffect(() => {
+    if (containerRef.current) {
+      containerRef.current.measure((x, y, width, height, pageX, pageY) => {
+        setContainerPosition({
+          x: pageX,
+          y: pageY,
+        });
+      });
+    }
+  }, [containerRef]);
+
+  const [anchorMeasure, setAnchorMeasure] = useState<Size | null>(null);
+  useEffect(() => {
+    if (anchorRef.current) {
+      anchorRef.current.measure((x, y, width, height) => {
+        setAnchorMeasure({
+          width,
+          height,
+        });
+      });
+    }
+  }, [anchorRef]);
+
+  useEffect(() => {
+    if (anchorMeasure?.width && tooltipRef.current) {
+      tooltipRef.current.measure((x, y, width, height) => {
+        dispatch({
+          type: "init",
+          payload: {
+            placement,
+            anchor: anchorMeasure,
+            tooltip: { width, height },
+          },
+        });
+      });
+    }
+  }, [anchorMeasure, show, placement]);
+
+  const trackCursor = useCallback(
+    (e: React.MouseEvent) => {
+      if (!containerPosition) return;
+      dispatch({
+        type: "move",
+        payload: {
+          top: e.clientY - containerPosition.y,
+          left: e.clientX - containerPosition.x,
```

Comment:
Hm, aren't we calculating the difference between a viewport coordinate and a document coordinate here? Isn't this only correct if the document is not scrolled?